// @convex-dev/rag/dist/component/embeddings/tables.js
import { literals } from "convex-helpers/validators";
import { defineTable, } from "convex/server";
import { v, } from "convex/values";
import { vectorWithImportanceDimension } from "./importance.js";
import { allFilterFieldNames, vAllFilterFields } from "../filters.js";
// We only generate embeddings for non-tool, non-system messages
const embeddingsFields = {
    vector: v.array(v.number()),
    ...vAllFilterFields,
};
function table(dimensions) {
    return defineTable(embeddingsFields).vectorIndex("vector", {
        vectorField: "vector",
        dimensions: vectorWithImportanceDimension(dimensions),
        filterFields: allFilterFieldNames,
    });
}
export const VectorDimensions = [
    128, 256, 512, 768, 1024, 1408, 1536, 2048, 3072, 4096,
];
export function assertVectorDimension(dimension) {
    if (!VectorDimensions.includes(dimension)) {
        throw new Error(`Unsupported vector dimension${dimension}. Supported: ${VectorDimensions.join(", ")}`);
    }
}
export function validateVectorDimension(dimension) {
    if (!VectorDimensions.includes(dimension)) {
        throw new Error(`Unsupported vector dimension${dimension}. Supported: ${VectorDimensions.join(", ")}`);
    }
    return dimension;
}
export const VectorTableNames = VectorDimensions.map((d) => `vectors_${d}`);
export const vVectorDimension = literals(...VectorDimensions);
export const vVectorTableName = literals(...VectorTableNames);
export const vVectorId = v.union(...VectorTableNames.map((name) => v.id(name)));
export function getVectorTableName(dimension) {
    return `vectors_${dimension}`;
}
// export function getVectorIdInfo(ctx: QueryCtx, id: VectorTableId) {
//   for (const dimension of VectorDimensions) {
//     const tableName = getVectorTableName(dimension);
//     if (ctx.db.normalizeId(tableName, id)) {
//       return { tableName, dimension };
//     }
//   }
//   throw new Error(`Unknown vector table id: ${id}`);
// }
const tables = Object.fromEntries(VectorDimensions.map((dimensions) => [
    `vectors_${dimensions}`,
    table(dimensions),
]));
export default tables;
