// @convex-dev/rag/src/component/_generated/api.ts
/* eslint-disable */
/**
 * Generated `api` utility.
 *
 * THIS CODE IS AUTOMATICALLY GENERATED.
 *
 * To regenerate, run `npx convex dev`.
 * @module
 */

import type * as chunks from "../chunks.js";
import type * as embeddings_importance from "../embeddings/importance.js";
import type * as embeddings_index from "../embeddings/index.js";
import type * as embeddings_tables from "../embeddings/tables.js";
import type * as entries from "../entries.js";
import type * as filters from "../filters.js";
import type * as namespaces from "../namespaces.js";
import type * as search from "../search.js";

import type {
  ApiFromModules,
  FilterApi,
  FunctionReference,
} from "convex/server";
import { anyApi, componentsGeneric } from "convex/server";

const fullApi: ApiFromModules<{
  chunks: typeof chunks;
  "embeddings/importance": typeof embeddings_importance;
  "embeddings/index": typeof embeddings_index;
  "embeddings/tables": typeof embeddings_tables;
  entries: typeof entries;
  filters: typeof filters;
  namespaces: typeof namespaces;
  search: typeof search;
}> = anyApi as any;

/**
 * A utility for referencing Convex functions in your app's public API.
 *
 * Usage:
 * ```js
 * const myFunctionReference = api.myModule.myFunction;
 * ```
 */
export const api: FilterApi<
  typeof fullApi,
  FunctionReference<any, "public">
> = anyApi as any;

/**
 * A utility for referencing Convex functions in your app's internal API.
 *
 * Usage:
 * ```js
 * const myFunctionReference = internal.myModule.myFunction;
 * ```
 */
export const internal: FilterApi<
  typeof fullApi,
  FunctionReference<any, "internal">
> = anyApi as any;

export const components = componentsGeneric() as unknown as {
  workpool: import("@convex-dev/workpool/_generated/component.js").ComponentApi<"workpool">;
};
