// @convex-dev/rag/dist/component/filters.js
import { v, } from "convex/values";
export const vFilterFieldValue = v.array(v.any());
export const filterFieldNames = [
    "filter0",
    "filter1",
    "filter2",
    "filter3",
];
export const vAllFilterFields = {
    namespaceId: v.id("namespaces"),
    filter0: v.optional(vFilterFieldValue),
    filter1: v.optional(vFilterFieldValue),
    filter2: v.optional(vFilterFieldValue),
    filter3: v.optional(vFilterFieldValue),
};
export const allFilterFieldNames = [
    "namespaceId",
    ...filterFieldNames,
];
export const vNamedFilter = v.object({
    name: v.string(),
    value: v.any(),
});
/**
 * { 1: "foo", 2: "bar" }
 *   -> { filter1: ["namespace", "foo"], filter2: ["namespace", "bar"] }
 */
export function filterFieldsFromNumbers(namespaceId, filters) {
    const filterFields = {};
    if (!filters)
        return filterFields;
    for (const [i, filter] of Object.entries(filters)) {
        const index = Number(i);
        if (isNaN(index) || index < 0 || index >= filterFieldNames.length) {
            console.warn(`Unknown filter index: ${index} for value ${JSON.stringify(filter)}`);
            break;
        }
        filterFields[filterFieldNames[index]] = [namespaceId, filter];
    }
    return filterFields;
}
/**
 * [{ name: "Foo", value: "foo" }, { name: "Baz", value: "baz" }]
 *   -> { 0: "foo", 2: "baz" }
 */
export function numberedFilterFromNamedFilters(namedFilters, filterNames) {
    const numberedFilter = {};
    for (const namedFilter of namedFilters) {
        const index = filterNames.indexOf(namedFilter.name);
        if (index === -1) {
            throw new Error(`Unknown filter name: ${namedFilter.name} for namespace with names ${filterNames.join(", ")}`);
        }
        numberedFilter[index] = namedFilter.value;
    }
    return numberedFilter;
}
/**
 * [{ name: "Foo", value: "foo" }, { name: "Baz", value: "baz" }]
 *   -> [{ 0: "foo" }, { 2: "baz" }]
 */
export function numberedFiltersFromNamedFilters(filters, filterNames) {
    const filterFields = [];
    for (const filter of filters) {
        const index = filterNames.indexOf(filter.name);
        if (index === -1) {
            throw new Error(`Unknown filter name: ${filter.name} for namespace with names ${filterNames.join(", ")}`);
        }
        filterFields.push({ [index]: filter.value });
    }
    return filterFields;
}
